packages/fxa-auth-server/lib/payments/stripe.ts

```diff
@@ -2299,6 +2327,68 @@ export class StripeHelper {
     return this.stripe[resourceType].retrieve(resource);
   }
 
+  async processWebhookEventToFirestore(event: Stripe.Event) {
+    const { type, data } = event;
+
+    if (!this.stripeFirestore) {
+      return;
+    }
+
+    // Note that we must insert before any event handled by the general
+    // webhook code to ensure the object is up to date in Firestore before
+    // our code handles the event.
+    try {
+      switch (type as Stripe.WebhookEndpointUpdateParams.EnabledEvent) {
+        case 'invoice.created':
+        case 'invoice.finalized':
+        case 'invoice.paid':
+        case 'invoice.payment_failed':
+        case 'invoice.updated':
+        case 'invoice.deleted':
+          const invoice = data.object as Stripe.Invoice;
+          await this.stripeFirestore.retrieveAndFetchSubscription(
+            invoice.subscription as string
+          );
+          // Now that the subscription is fetched, we can safely insert the
+          // invoice record.
+          await this.stripeFirestore.insertInvoiceRecord(invoice);
```

Comment:
Why are we doing this `retrieveAndFetchSubscription`?

This method will add the customer record and the subscription in Firestore if it doesn't exist, so I guess we're making sure there's a customer object with the associated subscription object loaded first, since the invoice collection is nested under the subscription collection (which is nested under customer)?

Reply:
We want the parent path to exist first before inserting invoices. This is also key to ensuring we back-fill active customer data pushes.